admin/controller/extension/module/altapay.php

```diff
@@ -20,180 +21,255 @@ public function index()
-                            chmod($language_path, 0664);
-                        } else {
-                            $data['refresh_terminals_error'] = $this->language->get('text_altapay_terminals_error');
-                        }
-
-                        $view_template_content = str_replace(array('{key}', '{_key_}', '{name}'), array($term_key, $term_key_, $term_name), $payment_template_view_template);
-                        $view_template_dir     = $dir . '/admin/view/template/extension/payment';
-
-                        // Check if language folder is writable
-                        if (is_writable($view_template_dir)) {
-                            file_put_contents($view_template_path, $view_template_content);
-                            chmod($view_template_path, 0664);
-                        } else {
-                            $data['refresh_terminals_error'] = $this->language->get('text_altapay_terminals_error');
-                        }
-
-                        $order_view_template_content = str_replace(array('{key}', '{_key_}', '{name}'), array($term_key, $term_key_, $term_name), $payment_template_order_view_template);
-                        $order_view_template_dir     = $dir . '/admin/view/template/extension/payment';
-
-                        // Check if language folder is writable
-                        if (is_writable($order_view_template_dir)) {
-                            file_put_contents($order_view_template_path, $order_view_template_content);
-                            chmod($order_view_template_path, 0664);
-                        } else {
-                            $data['refresh_terminals_error'] = $this->language->get('text_altapay_terminals_error');
-                        }
-
-                        // FRONTEND Templates
-                        $frontend_model_content = str_replace(array('{key}', '{_key_}', '{name}'), array($term_key, $term_key_, $term_name), $frontend_payment_model);
-                        $frontend_model_dir     = $dir . '/catalog/model/extension/payment';
-                        // Check if language folder is writable
-                        if (is_writable($frontend_model_dir)) {
-                            file_put_contents($frontend_model_path, $frontend_model_content);
-                            chmod($frontend_model_path, 0664);
-                        } else {
-                            $data['refresh_terminals_error'] = $this->language->get('text_altapay_terminals_error');
+        $data['refresh_terminals_error'] = '';
+        $data['sync_terminals_success'] = '';
+        $data['sync_terminals_error'] = '';
+
+        if (($this->request->server['REQUEST_METHOD'] == 'POST') && $this->validate() && isset($_POST['terminalsync'])) {
+
+            $terminals_refreshed_query = $this->db->query("SELECT `value` FROM " . DB_PREFIX . "setting WHERE `key` LIKE 'altapay_terminals_refreshed'");
```

Comment:
Always use prepared statements for querying, never concatenating strings

Reply:
[It looks like](https://github.com/opencart/opencart/blob/master/upload/system/library/db.php) OpenCart does not have support for prepared statement in their custom DB interface.

Reply:
@emicha I've taken measures to the maximum extent possible in OpenCart in the below commit belonging to the next pull request
https://github.com/AltaPay/plugin-opencart3/pull/17/commits/1256ccf8696af93c2b5e3784e4523d5b832e02bf